src/v/raft/id_allocator_stm.h

```diff
@@ -0,0 +1,174 @@
+// last component is persisted in RAM. All the 3 components guarantees
+// uniqueness even if the node is restarted (last component is reset but
+// the first is incremented).
+//
+// Log garbage collection. If we do nothing all the "+1000" commands will
+// eventually consume all the disk. To prevent this from happening
+// we use the following snapshotting scheme:
+//
+//   1. once the log size passes a threshold a leader writes
+//      "prepare_truncation" message to mark a place in the log
+//      where we want to cut off the history
+//
+//   2. when the leader catches up with the "prepare_truncation"
+//      it knows the current state (sum of all the messages before
+//      the mark) and the prepare_truncation's  so it writes
+//      "execute_truncation(state, offset)"
+//
+//   3. after the leader catches up with the execute_truncation
+//      command it truncates the local log before the mark by
+//      calling raft::write_snapshot
+//
+//   4. after a follower catches up with the execute_truncation
+//      it truncates its log
+//
+// There may be regular "+1000" commands between the "prepare" and the
+// "execute" commands. When this happen the state_machine postpone the
+// execution of the "+1000" commands until it meets the execute_truncation
+// and use its state as a base for the increments.
+// A leader starts acting before it reads up to the current moment
+// so there may be a situation when a leader issues several
+// "prepare_truncation" before it issues "execute_truncation" in this
+// case the first "prepare_truncation" in the log order wins. The same
+// works with "execute_truncation" the first wins.
+
+class id_allocator_stm final : public raft::state_machine {
+public:
+    struct stm_allocation_result {
+        int64_t id;
+        raft::errc raft_status{raft::errc::success};
+    };
+
+    explicit id_allocator_stm(
+      ss::logger&, raft::consensus*, config::configuration&);
+
+    ss::future<> start() final;
+
+    ss::future<stm_allocation_result>
+    allocate_id_and_wait(model::timeout_clock::time_point timeout);
+
+private:
+    struct sequence_id {
+        model::run_id run_id;
+        model::node_id node_id;
+        int64_t tick;
+
+        bool operator==(const sequence_id& other) const {
+            return run_id == other.run_id && node_id == other.node_id
+                   && tick == other.tick;
+        }
+
+        bool operator!=(const sequence_id& other) const {
+            return !(*this == other);
+        }
+
+        template<typename H>
+        friend H AbslHashValue(H h, const sequence_id& seq) {
+            return H::combine(std::move(h), seq.run_id, seq.node_id, seq.tick);
+        }
+    };
+
+    struct log_allocation_result {
+        sequence_id seq;
+        int64_t base;
+        int64_t range;
+        raft::errc raft_status{raft::errc::success};
+        model::offset offset;
+    };
+
+    struct allocation_cmd {
```

Comment:
Since it looks like we're going to merge this PR before we finish the raft snapshots that would have simplified things, is there any additional state that should be stored in the log that would be needed for a future migration to the to-be-added raft snapshot feature?

The obvious candidate would be to add a `int8_t` version as the first element to each of the messages that are serialized into the log and setting that value to 0 in this PR. Anything being stored to disk should generally have such a version. But the change needed for the future migration might be a version in the snapshot state.

Anyway, we should make sure we future proof it which shouldn't be more complex than making sure we are storing a version in the right location. Maybe it's we already have enough state?

Reply:
Discussed offline, it's ok to leave it this way for a while. Going to switch to raft's snapshots before it goes live.